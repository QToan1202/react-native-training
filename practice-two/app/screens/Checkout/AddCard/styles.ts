import { Dimensions, StyleSheet } from 'react-native'

import { COLORS } from '@constants'

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: COLORS.BG_LAYER,
  },
  form: {
    flex: 1,
    backgroundColor: COLORS.WHITE,
    paddingVertical: 24,
    paddingHorizontal: 34,
  },
  cardContainer: {
    width: Dimensions.get('window').width,
    height: Dimensions.get('window').width / 2,
    backgroundColor: COLORS.BG_LAYER,
    paddingHorizontal: 34,
  },
  cardSize: {
    width: 310,
    height: 183,
    alignSelf: 'center',
    position: 'absolute',
    top: -15,
    zIndex: 3,
  },
  cardContent: {
    maxHeight: undefined,
    maxWidth: undefined,
    flex: 1,
    alignSelf: 'stretch',
    justifyContent: 'space-between',
    paddingRight: 0,
  },
  expiresDate: {
    flex: 3,
  },
  cvc: {
    flex: 2,
  },
  spacing: {
    columnGap: 30,
  },
  error: {
    paddingLeft: 0,
  },
})

export default styles
