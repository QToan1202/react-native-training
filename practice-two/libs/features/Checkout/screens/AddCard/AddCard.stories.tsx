import { ComponentMeta, ComponentStory } from '@storybook/react-native'

import { AddCard } from '../index'

export default {
  title: 'screens/Checkout/Add Card',
  component: AddCard,
} as ComponentMeta<typeof AddCard>

const Template: ComponentStory<typeof AddCard> = () => <AddCard />

export const Default = Template.bind({})
Default.args = {}
