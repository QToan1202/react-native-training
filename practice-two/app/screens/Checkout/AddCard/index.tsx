import { useCallback } from 'react'
import { Dimensions, KeyboardAvoidingView, Modal, Platform } from 'react-native'
import { useForm } from 'react-hook-form'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import { ScrollView, XStack, YStack } from 'tamagui'
import { useShallow } from 'zustand/react/shallow'

import { RootStackParamList } from '@navigation/Stack'
import { Bar, Input, PaymentCard, TabBar } from '@components'
import { ICardBase, IForm } from '@types'
import { useAddCard } from '@hooks'
import { useAuthStore } from '@stores'

import styles from './styles'

export type AddCardScreenProps = NativeStackScreenProps<RootStackParamList, 'AddCard'>
const AddCard = ({ navigation }: AddCardScreenProps) => {
  const [isHydrated, user] = useAuthStore(useShallow((state) => [state.isHydrated, state.user]))
  const { mutate } = useAddCard(process.env.CARD_ENDPOINT)
  const { control, formState, watch, handleSubmit } = useForm<IForm>()
  const observeFields = watch()
  const handleSaveCardInfo = useCallback(
    (data: ICardBase) => {
      // TODO: Save card information to context
      if (!isHydrated) return
      if (!user) return

      mutate(
        {
          ...data,
          userId: user.id,
        },
        {
          onSuccess: () => {
            navigation.goBack()
          },
        }
      )
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [isHydrated, user]
  )
  const handleGoBack = useCallback(() => navigation.goBack(), [navigation])

  return (
    <>
      <Bar
        title="add card"
        align="center"
        paddingBottom="$space.6"
        showBackBtn
        onPressBack={handleGoBack}
      />
      <XStack
        justifyContent="center"
        backgroundColor="$color.bg_layer"
        width={Dimensions.get('window').width}
        height={Dimensions.get('window').width / 2}
      >
        <PaymentCard
          name={observeFields.name || ''}
          number={observeFields.number || ''}
          expired={observeFields.expired || ''}
          cvc={observeFields.cvc || ''}
          alignSelf="center"
          marginBottom={30}
          pressStyle={{ opacity: 1 }}
          top="$space.-4"
          cardStyle={{
            maxWidth: undefined,
            maxHeight: undefined,
            aspectRatio: 3 / 1.8,
          }}
        />
      </XStack>
      <ScrollView backgroundColor="$color.white">
        <KeyboardAvoidingView
          behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
          style={styles.form}
        >
          <YStack space>
            <Input
              name="number"
              label="Card Number"
              control={control}
              keyboardType="numeric"
              isShowError
              rules={{
                required: 'Card number is require',
                maxLength: {
                  value: 16,
                  message: "Card number can't be exceed 16 number",
                },
                validate: (value) => value[0] === '4' || value[0] === '5' || 'Card number invalid',
              }}
            />
            <Input
              name="name"
              control={control}
              label="Name"
              rules={{
                required: 'Name is require',
              }}
            />
            <XStack alignItems="center" space="$space.6">
              <Input
                label="Expires Dates"
                name="expired"
                isShowError
                control={control}
                rules={{
                  required: 'Expires Date is require',
                }}
                containerStyle={{ flex: 3 }}
              />
              <Input
                label="CVC"
                name="cvc"
                control={control}
                isShowError
                keyboardType="numeric"
                secureTextEntry
                rules={{
                  required: 'CVC is require',
                  maxLength: {
                    value: 3,
                    message: 'CVC too long',
                  },
                }}
                containerStyle={{ flex: 2 }}
              />
            </XStack>
          </YStack>
        </KeyboardAvoidingView>
      </ScrollView>
      <TabBar
        title="Add Credit Card"
        isDisable={!!Object.keys(formState.errors).length}
        onPress={handleSubmit(handleSaveCardInfo)}
      />
    </>
  )
}

export default AddCard
