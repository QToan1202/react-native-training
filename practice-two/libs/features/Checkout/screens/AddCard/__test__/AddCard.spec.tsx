import { render } from '@testing-library/react-native'

import AddCard from '../index'

describe('Testing Add Card screen', () => {
  it('Render screen successfully', () => {
    const component = render(<AddCard />).toJSON()

    expect(component).toMatchSnapshot()
  })
})
