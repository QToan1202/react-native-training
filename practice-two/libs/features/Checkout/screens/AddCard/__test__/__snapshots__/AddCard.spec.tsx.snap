// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Add Card screen Render screen successfully 1`] = `
[
  <View
    style={
      {
        "backgroundColor": "#f6f9ff",
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#f6f9ff",
          "height": 375,
          "paddingHorizontal": 34,
          "width": 750,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignSelf": "center",
            "height": 183,
            "opacity": 1,
            "position": "absolute",
            "top": -15,
            "width": 310,
            "zIndex": 3,
          }
        }
      >
        <View
          accessibilityIgnoresInvertColors={true}
          style={
            [
              {
                "alignItems": "flex-end",
                "alignSelf": "flex-start",
                "borderRadius": 9,
                "columnGap": 17,
                "flexDirection": "row",
                "maxHeight": 146,
                "maxWidth": 246,
                "overflow": "hidden",
                "paddingHorizontal": 17,
                "paddingVertical": 13,
              },
              {
                "alignSelf": "stretch",
                "flex": 1,
                "justifyContent": "space-between",
                "maxHeight": undefined,
                "maxWidth": undefined,
                "paddingRight": 0,
              },
            ]
          }
        >
          <Image
            source={
              {
                "testUri": "../../../assets/payment/normal.png",
              }
            }
            style={
              [
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                {
                  "height": undefined,
                  "width": undefined,
                },
                undefined,
              ]
            }
          />
          <View
            style={
              {
                "height": "100%",
                "justifyContent": "space-between",
              }
            }
          >
            <View>
              <Text
                style={
                  [
                    {
                      "color": "#fff",
                      "fontFamily": "Montserrat_400Regular",
                    },
                    {
                      "fontSize": 14,
                    },
                    {
                      "color": "#b7b7b7",
                      "fontSize": 13,
                      "textTransform": "capitalize",
                    },
                  ]
                }
              >
                holder name
              </Text>
              <Text
                numberOfLines={1}
                style={
                  [
                    {
                      "color": "#fff",
                      "fontFamily": "Montserrat_400Regular",
                    },
                    {
                      "fontSize": 14,
                    },
                    undefined,
                  ]
                }
              />
            </View>
            <View>
              <Text
                style={
                  [
                    {
                      "color": "#fff",
                      "fontFamily": "Montserrat_400Regular",
                    },
                    {
                      "fontSize": 14,
                    },
                    {
                      "color": "#b7b7b7",
                      "fontSize": 13,
                      "textTransform": "capitalize",
                    },
                  ]
                }
              >
                card number
              </Text>
              <Text
                numberOfLines={1}
                style={
                  [
                    {
                      "color": "#fff",
                      "fontFamily": "Montserrat_400Regular",
                    },
                    {
                      "fontSize": 16,
                    },
                    {
                      "width": 167,
                    },
                  ]
                }
              />
            </View>
            <View>
              <Text
                style={
                  [
                    {
                      "color": "#fff",
                      "fontFamily": "Montserrat_400Regular",
                    },
                    {
                      "fontSize": 14,
                    },
                    {
                      "color": "#b7b7b7",
                      "fontSize": 13,
                      "textTransform": "capitalize",
                    },
                  ]
                }
              >
                exp. date
              </Text>
              <Text
                style={
                  [
                    {
                      "color": "#fff",
                      "fontFamily": "Montserrat_400Regular",
                    },
                    {
                      "fontSize": 14,
                    },
                    undefined,
                  ]
                }
              />
            </View>
          </View>
          <View>
            <Text
              style={
                [
                  {
                    "color": "#fff",
                    "fontFamily": "Montserrat_400Regular",
                  },
                  {
                    "fontSize": 14,
                  },
                  [
                    {
                      "color": "#b7b7b7",
                      "fontSize": 13,
                      "textTransform": "capitalize",
                    },
                    {
                      "textTransform": "uppercase",
                    },
                  ],
                ]
              }
            >
              cvc
            </Text>
            <Text
              style={
                [
                  {
                    "color": "#fff",
                    "fontFamily": "Montserrat_400Regular",
                  },
                  {
                    "fontSize": 14,
                  },
                  {
                    "textAlign": "right",
                  },
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        [
          {
            "backgroundColor": "#fff",
            "flex": 1,
            "paddingHorizontal": 34,
            "paddingTop": 24,
          },
          {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View>
        <View>
          <Text
            style={
              [
                {
                  "color": "#fff",
                  "fontFamily": "Montserrat_400Regular",
                },
                {
                  "fontSize": 14,
                },
                {
                  "color": "#3a2c3c",
                  "fontFamily": "Montserrat_500Medium",
                },
              ]
            }
          >
            Card Number
          </Text>
          <TextInput
            keyboardType="numeric"
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            style={
              [
                {
                  "borderBottomColor": "#dbdbde",
                  "borderBottomWidth": 0.5,
                  "color": "#000",
                  "fontFamily": "Montserrat_500Medium",
                  "fontSize": 16,
                  "paddingVertical": 5,
                },
                {},
                undefined,
              ]
            }
            value=""
          />
        </View>
        <View
          style={
            {
              "height": 20,
            }
          }
        />
      </View>
      <View>
        <View>
          <Text
            style={
              [
                {
                  "color": "#fff",
                  "fontFamily": "Montserrat_400Regular",
                },
                {
                  "fontSize": 14,
                },
                {
                  "color": "#3a2c3c",
                  "fontFamily": "Montserrat_500Medium",
                },
              ]
            }
          >
            Name
          </Text>
          <TextInput
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            style={
              [
                {
                  "borderBottomColor": "#dbdbde",
                  "borderBottomWidth": 0.5,
                  "color": "#000",
                  "fontFamily": "Montserrat_500Medium",
                  "fontSize": 16,
                  "paddingVertical": 5,
                },
                {},
                undefined,
              ]
            }
            value=""
          />
        </View>
        <View
          style={
            {
              "height": 20,
            }
          }
        />
      </View>
      <View
        style={
          [
            {
              "alignItems": "center",
              "columnGap": 4,
              "flexDirection": "row",
            },
            {
              "columnGap": 30,
            },
          ]
        }
      >
        <View
          style={
            {
              "flex": 3,
            }
          }
        >
          <View>
            <Text
              style={
                [
                  {
                    "color": "#fff",
                    "fontFamily": "Montserrat_400Regular",
                  },
                  {
                    "fontSize": 14,
                  },
                  {
                    "color": "#3a2c3c",
                    "fontFamily": "Montserrat_500Medium",
                  },
                ]
              }
            >
              Expires Dates
            </Text>
            <TextInput
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              style={
                [
                  {
                    "borderBottomColor": "#dbdbde",
                    "borderBottomWidth": 0.5,
                    "color": "#000",
                    "fontFamily": "Montserrat_500Medium",
                    "fontSize": 16,
                    "paddingVertical": 5,
                  },
                  {},
                  undefined,
                ]
              }
              value=""
            />
          </View>
          <View
            style={
              {
                "height": 20,
              }
            }
          />
        </View>
        <View
          style={
            {
              "flex": 2,
            }
          }
        >
          <View>
            <Text
              style={
                [
                  {
                    "color": "#fff",
                    "fontFamily": "Montserrat_400Regular",
                  },
                  {
                    "fontSize": 14,
                  },
                  {
                    "color": "#3a2c3c",
                    "fontFamily": "Montserrat_500Medium",
                  },
                ]
              }
            >
              CVC
            </Text>
            <TextInput
              keyboardType="numeric"
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              secureTextEntry={true}
              style={
                [
                  {
                    "borderBottomColor": "#dbdbde",
                    "borderBottomWidth": 0.5,
                    "color": "#000",
                    "fontFamily": "Montserrat_500Medium",
                    "fontSize": 16,
                    "paddingVertical": 5,
                  },
                  {},
                  undefined,
                ]
              }
              value=""
            />
          </View>
          <View
            style={
              {
                "height": 20,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>,
  <View
    style={
      [
        {
          "paddingHorizontal": 16,
          "paddingVertical": 22,
        },
        {
          "backgroundColor": "#33907C",
          "rowGap": 24,
        },
        {
          "paddingTop": 0,
        },
        [
          {
            "backgroundColor": "#fff",
            "paddingHorizontal": 32,
          },
          {
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 12,
              "width": 0,
            },
            "shadowOpacity": 0.6,
            "shadowRadius": 16,
          },
          {
            "paddingBottom": 12,
            "paddingTop": 12,
          },
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "columnGap": 4,
            "flexDirection": "row",
          },
          {
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#33907C",
            "borderColor": "#33907C",
            "borderRadius": 50,
            "borderWidth": 2,
            "flex": 1,
            "opacity": 1,
            "paddingVertical": 13,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "columnGap": 6,
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "#fff",
                  "fontFamily": "Montserrat_400Regular",
                },
                {
                  "fontSize": 16,
                },
                [
                  {
                    "color": "#fff",
                  },
                  {
                    "fontFamily": "Montserrat_500Medium",
                    "textAlign": "center",
                    "textTransform": "capitalize",
                  },
                  undefined,
                ],
              ]
            }
          >
            Add Credit Card
          </Text>
        </View>
      </View>
    </View>
  </View>,
]
`;
